### ISSUE TITLE
Fails to pin version of gcm services

### ISSUE THREAD

**Issue by zen0wu** [OneSignal/react-native-onesignal#445]
Apologize if this is the wrong place to ask, it might be problem with onesignal-android-sdk itself.

I added the following in build.gradle

```
dependencies {
    compile project(':react-native-onesignal')
    compile "com.google.android.gms:play-services-base:11.4.0"
}
```

It seems that react-native-onesignal's dependency declaration forces gradle to upgrade the gms version to 12.0.1, but I couldn't figure it out since the OneSignal 3.8.3's POM says it's ok with all versions between `[10.2.1, 12.1.0)`

http://central.maven.org/maven2/com/onesignal/OneSignal/3.8.3/OneSignal-3.8.3.pom

**Comment by jkasten2** [comment]
@shivawu OneSignal works with version `10.2.1` and newer and any version before `12.1.0` of gms. Gradle doesn't pin to a version, it will try to use the newest possible in that range.

By adding `compile com.google.android.gms:play-services-base:11.4.0` it won't have any effect as `play-services-gcm:12.0.1` has an exact version dependency on `play-services-base:12.0.1`.

See this output of `./gradlew app:dependencies` on a `build.gradle` like the one you noted above
```
+--- com.onesignal:OneSignal:[3.8.3, 3.99.99] -> 3.8.3
|    +--- com.google.android.gms:play-services-gcm:[10.2.1, 12.1.0) -> 12.0.1
|    |    +--- com.google.android.gms:play-services-base:12.0.1
...
|
+--- com.google.android.gms:play-services-base:11.4.0 -> 12.0.1 (*)
```

The general rule with `com.google.android.gms` is all modules under it `play-services-gcm`, `play-services-location`, etc must all have the exact same version. Otherwise there is either build or runtime issues.

Is there is any specific reason you need to down grade to an older version of `com.google.android.gms`?

**Comment by zen0wu** [comment]
Thanks a lot for your kind answer!

There's actually no specific reason that I have to pin the version, but I found every time upstream's gms dependency has upgrade it would break my build. Then I found the issue is because some of my 3rd-party dependencies (react-native-device-info, react-native-onesignal for example) have specific version pinned in their build, so I have to cope with the same version.

Now many of them seem to have to move to a best practice that having a global config to specify the version from the root project (https://github.com/rebeccahughes/react-native-device-info/blob/master/android/build.gradle#L27), the only dependency left that seems to have an effect on my specific version is react-native-onesignal -> com.onesignal:OneSignal, so I'm trying to understand how it works.

What you're saying is, specifying a version in my build file wouldn't do any thing, since gradle:
1. Looks at the dependency com.onesignal:OneSignal
2. Looks at its dependency, com.google.android.gms:play-services-gcm:[10.2.1, 12.1.0)
3. Choose `12.0.1` among them, since it's the latest
4. Looks at the next dependency `com.google.android.gms:play-services-base:11.4.0`
5. Resolve the conflict between `12.0.1` and `11.4.0`, which results in `12.0.1`

So essentially there's no way that we could affect this resolution process, meaning I have to go with what my 3rd-party dependency A (in this case react-native-onesignal -> com.onesignal:OneSignal), and pick their versions, in my `googlePlayServicesVersion` (say I use the way that these projects work with these versions)?

It means if I have another dependency B, which specify a different version of gcm, I could be in trouble?

**Comment by jkasten2** [comment]
@shivawu Sorry for the long delay, been busy with a number of projects!

I agree that defining a global versions for dependencies such as `project.googlePlayServicesVersion` takes care of a lot of the headache but is still is a point of maintenance and failure. Some plugins only work with certain versions of gms, firebase, android support etc so you could be downgrading or upgrading to far. So `project.googlePlayServicesVersion` would have to be updated every so often to fix either compile or runtime errors when updating or when adding new plugins / SDKs to your project. Basically a global like this is force overriding versions! Instead a version should be pick automatically at build time considering all plugins and their dependent versions.

So Gradle doesn't do the alignment of groups that is needed to automatically resolve a "best" version. It ends up leaving different modules of `com.google.android.gms` mismatched which results in either compile or runtime errors. Example: One plugin may set `com.google.android.gms:play-services-gcm:11.4.0` and another may set `com.google.android.gms:play-services-maps:9.8.0` which will result in a compile or runtime error in most cases. This is a problem due to a mixture of interdependencies as `play-services-maps:9.8.0` might try to use something on `play-services-base:11.4.0` but it was removed in `play-services-base:11.0.0`. Just an example, these 2 version might not actually be a problem.

To fix this issue I created a Gradle plugin that will inspect all dependences in the project and align them to the best possible version automatically! I just released a 0.9.0 today that does version pining as you noted, even though sub projects.
https://github.com/OneSignal/OneSignal-Gradle-Plugin

Let me know if this solves the kind issues you have been facing.
